src/main/java/org/broadinstitute/consent/http/service/UserService.java

```diff
@@ -454,12 +454,135 @@ public void hasValidActiveERACredentials(User user) {
     }
   }
 
+  /**
+   * Compliance method defined in
+   * <a href="https://broadworkbench.atlassian.net/browse/DT-1607">this ticket</a> that implements
+   * a set of rules in order to ensure Library Card and Institution matching rules are
+   * adhered to when authorizing users of the system.
+   * @param user the User being evaluated
+   * @return user with the Institution and Library Card rules applied
+   */
+  public User enforceInstitutionAndLibraryCardRules(User user) {
+    Institution institutionFromEmail = institutionService.findInstitutionForEmail(user.getEmail());
+    Institution institutionFromDatabase = null;
+    try {
+      institutionFromDatabase = institutionService.findInstitutionById(user.getInstitutionId());
+    } catch (NotFoundException nfe) {
+      // do nothing.
+    }
+
+    boolean hasInstitutionMatchingEmailDomain =
+        hasInstitutionMatchingEmailDomain(institutionFromEmail);
+
+    boolean modifiedUser = false;
+
+    if (hasInstitutionMatchingEmailDomain) {
+      if (handleUserWithInstitutionInMap(user, institutionFromEmail, institutionFromDatabase)) {
+        modifiedUser = true;
+      }
+    } else {
+      if (handleUserWithoutInstitutionInMap(user, institutionFromDatabase)) {
+        modifiedUser = true;
+      }
+    }
+
+    if (modifiedUser) {
+      return findUserByEmail(user.getEmail());
+    } else {
+      return user;
+    }
+  }
+
+  @VisibleForTesting
+  protected boolean handleUserWithInstitutionInMap(User user, Institution institutionFromEmail, Institution institutionFromDatabase) {
+    boolean modifiedUser = false;
+    if (!hasMatchingInstitutionInDatabase(institutionFromEmail, institutionFromDatabase)) {
+      assignInstitutionToUser(user, institutionFromEmail.getId());
+      // Set these so we don't have to look the user up again in the database for the next bit of
+      // work with the library card.
+      user.setInstitutionId(institutionFromEmail.getId());
+      user.setInstitution(institutionFromEmail);
+      modifiedUser = true;
+    }
+
+    if (handleLibraryCardForUser(user)) {
+      modifiedUser = true;
+    }
+
+    return modifiedUser;
+  }
+
+  @VisibleForTesting
+  protected boolean handleLibraryCardForUser(User user) {
+    boolean modifiedUser = false;
+    if (hasLibraryCard(user)) {
+      try {
+        User lcIssuer = findUserById(user.getLibraryCards().get(0).getCreateUserId());
+        Institution lcIssuerInstitution = institutionService.findInstitutionForEmail(lcIssuer.getEmail());
+        if (!user.getInstitution().equals(lcIssuerInstitution)) {
+          dropLibraryCardForUser(user);
+          modifiedUser = true;
+        }
+      } catch (NotFoundException nfe) {
+        dropLibraryCardForUser(user);
+        modifiedUser = true;
+      }
+    }
+    return modifiedUser;
+  }
+
+  @VisibleForTesting
+  protected boolean handleUserWithoutInstitutionInMap(User user, Institution institutionFromDatabase) {
+    if (hasLibraryCard(user)) {
```

Comment:
Need for this is reinforced by [https://broadworkbench.atlassian.net/browse/DT-1669](https://broadworkbench.atlassian.net/browse/DT-1669)